### ISSUE TITLE
Generics partially working on column names - but not values.

### ISSUE THREAD

**Issue by midwesterntechnology** [koskimas/kysely#43]
I've gotten generics partly working. Typescript will infer column names. However on values, it does not infer types.

For example, trying to do query.where("field", "=", "value"), the column name works correctly. It knows which columns are valid and which are not. However, for the value parameter, there is no data type that will work.

Here's some code which explains more clearly what works and what doesn't.

```ts
import { SelectQueryBuilder, UpdateQueryBuilder, UpdateResult } from "kysely";

type TestMeta = {
    isDirty: boolean
    hasBeenPersisted: boolean
    syncMessage: string
}
type SyncDef<DB extends { [key in K]: TestMeta }, K extends keyof DB> = {
    select: () => SelectQueryBuilder<DB, K, {}>
    update: () => UpdateQueryBuilder<DB, K, UpdateResult>
}

async function testSyncDef<DB extends { [key in K]: TestMeta }, K extends keyof DB>(def: SyncDef<DB, K>) {

    //test.isDirty is a boolean - GOOD!
    const test: DB[K] = undefined!;
    const dirty = test.isDirty;

    //these columns work - GOOD!
    def.select().whereRef("isDirty", "=", "hasBeenPersisted")

    //this gives an error on the non-existent column name - GOOD!
    def.select().where("doesNotExist", "=", "X");

    //using select
    //results are fine - GOOD!
    const isDirtyResults = await def.select().select(["isDirty", "hasBeenPersisted"]).execute();
    isDirtyResults[0].isDirty;
    isDirtyResults[0].hasBeenPersisted;

    //using selectAll
    //results are a promise of array of UnionToIntersection<DB[K]>
    //results does not have isdirty, even though DB[K] does? maybe because of UnionToIntersection type? - NOT GOOD
    const results = await def.select().selectAll().execute();
    results[0].isDirty;

    //Argument of type '{ isDirty: true; }' is not assignable to parameter of type 'MutationObject<DB, K>
    //column names are recognized, but it works when all values are undefined - no types work, even null (probably related to the below error)
    await def.update().set({
        isDirty: undefined,
        hasBeenPersisted: undefined,
        syncMessage: undefined,
        xxx: undefined,//gives an error for non-existent column - good!
    })

    //this gives an error on the "true" - last arg - not good - there is no data type that works here
    //Argument of type 'true' is not assignable to parameter of type 'FilterValueExpressionOrList<DB, K, "isDirty">'
    def.select().where("isDirty", "=", true);

}
```

**Comment by koskimas** [comment]
The problem here seems to be that you don't have any table names. The type for the table names in `DB` and `TB` arguments of Kysely are just `string` because that's the narrowest type typescript can be sure of based on `keyof DB`.

You have the same issue/misconception about typescript here as you did in the previous issue. Typescript doesn't carry around generic types in the code, but instead infers the narrowest concrete type in their place.

Since you have defined DB and K using a circular definition, the best typescript can do is infer that `K` must be something that can be a key of an object (string or more specifically keyof any) and that the values of `DB` all extend `TestMeta`.

Instead of a generic `DB`, why can't you just define the `DB` interface somewhere and use that? Why would you try to make this generic?

**Comment by koskimas** [comment]
Another problem: `K extends keyof DB` means that `K` has all the keys of `DB` BUT it may have more. Calling `DB[K]` will always be `never` because typescript cannot know if `K` is one of the keys of DB or one of the other keys.

**Comment by koskimas** [comment]
Actually you may be right about UnionToIntersect. This all might work without it. I'll take a look if it's possible to get rid of it. I'm pretty sure it's not, but let's see.

**Comment by midwesterntechnology** [comment]
>The problem here seems to be that you don't have any table names. The type for the table names in DB and TB arguments of Kysely are just string because that's the narrowest type typescript can be sure of based on keyof DB.

>Since you have defined DB and K using a circular definition, the best typescript can do is infer that K must be something that can be a key of an object (string or more specifically keyof any) and that the values of DB all extend TestMeta.

I'm probably missing something but this I don't quite follow. I'm no longer using strings to refer to tables. Each SyncDef returns a query builder of the correct type, a minor redundancy but no big deal. The column names and values are picked up properly on DB[K] itself - just Kysely can not pick up the value types, only names.

```
    const test: DB[K] = undefined!;
    const dirty = test.isDirty; //this is a boolean
    const isDirtyResults = await def.select().select(["isDirty"]).execute();
    isDirtyResults[0].isDirty; //this is a boolean
```

Typescript correctly infers that DB[K].isDirty is a boolean here. Kysely even returns the correct type.

Long story short, here's where I'm confused:

`Selection<DB, K, "isDirty">[]` correctly infers that isDirty is a boolean
But `FilterValueExpressionOrList<DB, K, "isDirty">` does not seem to be able to do the same.

>Instead of a generic DB, why can't you just define the DB interface somewhere and use that? Why would you try to make this generic?

Most of my code does this, but we have 20+ entities that we need to loop over in order to do a synchronization process with a master database, with quite a bit of generic logic. In addition, I've put a thin ORM layer over Kysely for updates/selects/inserts, which works fine, I just can't do anything that requires filtering through Kysely.

**Comment by koskimas** [comment]
```ts
    const test: DB[K] = undefined!;
    const dirty = test.isDirty; //this is a boolean
    const isDirtyResults = await def.select().select(["isDirty"]).execute();
    isDirtyResults[0].isDirty;
```

The type of the last line is `never`, not `boolean`

**Comment by midwesterntechnology** [comment]
> ```ts
>     const test: DB[K] = undefined!;
>     const dirty = test.isDirty; //this is a boolean
>     const isDirtyResults = await def.select().select(["isDirty"]).execute();
>     isDirtyResults[0].isDirty;
> ```
> 
> The type of the last line is `never`, not `boolean`

You're right, my mistake.

**Comment by koskimas** [comment]
This all works if you drop the generics by the way.

```ts
  interface DB {
    a: {
      isDirty: boolean
      hasBeenPersisted: boolean
      syncMessage: string
      foo: string
    }
    b: {
      isDirty: boolean
      hasBeenPersisted: boolean
      syncMessage: string
      bar: number
    }
    c: {
      isDirty: boolean
      hasBeenPersisted: boolean
      syncMessage: string
    }
  }

  type SyncDef<K extends keyof DB> = {
    select: () => SelectQueryBuilder<DB, K, {}>
    update: () => UpdateQueryBuilder<DB, K, UpdateResult>
  }

  async function testSyncDef(def: SyncDef<keyof DB>) {
    //these columns work - GOOD!
    def.select().whereRef('isDirty', '=', 'hasBeenPersisted')

    //this gives an error on the non-existent column name - GOOD!
    def.select().where('doesNotExist', '=', 'X')

    //using select
    //results are fine - GOOD!
    const isDirtyResults = await def
      .select()
      .select(['isDirty', 'hasBeenPersisted'])
      .execute()
    isDirtyResults[0].isDirty
    isDirtyResults[0].hasBeenPersisted

    //using selectAll
    //results are a promise of array of UnionToIntersection<DB[K]>
    //results does not have isdirty, even though DB[K] does? maybe because of UnionToIntersection type? - NOT GOOD
    const results = await def.select().selectAll().execute()
    results[0].isDirty

    //Argument of type '{ isDirty: true; }' is not assignable to parameter of type 'MutationObject<DB, K>
    //column names are recognized, but it works when all values are undefined - no types work, even null (probably related to the below error)
    await def.update().set({
      isDirty: false,
      hasBeenPersisted: undefined,
      syncMessage: undefined,
      xxx: undefined, //gives an error for non-existent column - good!
    })

    //this gives an error on the "true" - last arg - not good - there is no data type that works here
    //Argument of type 'true' is not assignable to parameter of type 'FilterValueExpressionOrList<DB, K, "isDirty">'
    def.select().where('isDirty', '=', true)
  }
```

**Comment by midwesterntechnology** [comment]
> This all works if you drop the generics by the way.
> 
> ```ts
>   interface DB {
>     a: {
>       isDirty: boolean
>       hasBeenPersisted: boolean
>       syncMessage: string
>       foo: string
>     }
>     b: {
>       isDirty: boolean
>       hasBeenPersisted: boolean
>       syncMessage: string
>       bar: number
>     }
>     c: {
>       isDirty: boolean
>       hasBeenPersisted: boolean
>       syncMessage: string
>     }
>   }
> 
>   type SyncDef<K extends keyof DB> = {
>     select: () => SelectQueryBuilder<DB, K, {}>
>     update: () => UpdateQueryBuilder<DB, K, UpdateResult>
>   }
> 
>   async function testSyncDef(def: SyncDef<keyof DB>) {
>     def.select().whereRef('isDirty', '=', 'hasBeenPersisted')
> 
>     def.select().where('doesNotExist', '=', 'X')
> 
>     const isDirtyResults = await def
>       .select()
>       .select(['isDirty', 'hasBeenPersisted'])
>       .execute()
>     isDirtyResults[0].isDirty
>     isDirtyResults[0].hasBeenPersisted
> 
>     const results = await def.select().selectAll().execute()
>     results[0].isDirty
> 
>     await def.update().set({
>       isDirty: false,
>       hasBeenPersisted: undefined,
>       syncMessage: undefined,
>       xxx: undefined, //gives an error for non-existent column - good!
>     })
> 
>     def.select().where('isDirty', '=', true)
>   }
> 
>   const a: SyncDef<'a'> = {} as unknown as any
>   const b: SyncDef<'b'> = {} as unknown as any
> 
>   testSyncDef(a)
>   testSyncDef(b)
> ```

Interesting. I just made this work (it did not work with my DB object before because two entities had the same named field, but with different types).

**Comment by koskimas** [comment]
It does work even if the entities have same fields with different types. The type of those fields is `never` though, which is actually correct. If you select the same column from two tables, you get an error from the db. So `never` kind of makes sense.

```ts
  interface DB {
    a: {
      isDirty: boolean
      hasBeenPersisted: boolean
      syncMessage: string
      foo: string
    }
    b: {
      isDirty: boolean
      hasBeenPersisted: boolean
      syncMessage: string
      bar: number
    }
    c: {
      isDirty: boolean
      hasBeenPersisted: boolean
      syncMessage: string
      bar: string
    }
  }
```

The above works as long as you don't select `bar`.

**Comment by midwesterntechnology** [comment]
> The above works as long as you don't select `bar`.

Strange, as soon as I changed those 2 fields it worked fine.

As far as using a specific DB object - it does work there in that file, however once I try to loop over entities (I am using continuations to mimic an existential type) - it's back to the same issue as before.

```
services.db.syncables.map(_ => _(async syncable => {
			await syncable.entity.access.select().where("isDirty", "=", false).execute();
}));
```

Gives me:

`Argument of type 'false' is not assignable to parameter of type 'FilterValueExpressionOrList<Schema, K, "isDirty">'`

Schema is my custom DB object (I am using it in the same manner as your example and it does work there).

Here is the type of syncable:

```
type ExSyncable = <R>(cont: <K extends keyof Schema>(prop: SyncDef<K>) => R) => R;

export function makeExSyncable<K extends keyof Schema>(property: SyncDef<K>): ExSyncable {
	return <R>(cont: <K extends keyof Schema>(prop: SyncDef<K>) => R) => cont(property);
}
```

Now, if I change this to use <keyof Schema> as you did, it fixes that issue but creates another because then I lose all specificity within my loop. For example, if I do this:

```
type ExSyncable = <R>(cont: (prop: SyncDef<keyof Schema>) => R) => R;

export function makeExSyncable(property: SyncDef<keyof Schema>): ExSyncable {
	return <R>(cont: (prop: SyncDef<keyof Schema>) => R) => cont(property);
}
```

Then access it within the loop:
```
const xxx = await syncable.entity.access.select().where("isDirty", "=", false).executeTakeFirstOrThrow();
```

I get a union of all my types, whereas before, I got this (notice i had to remove the where - or I would get the same error as before):

```
const xxx = await syncable.entity.access.select().selectAll().executeTakeFirstOrThrow();
```
xxx is of type `const xxx: Awaited<{} & UnionToIntersection<Schema[K]>>`

**Comment by midwesterntechnology** [comment]
I put together an example that works independently:

```
import { SelectQueryBuilder, UpdateQueryBuilder, UpdateResult } from "kysely";

type TestMeta = {
    isDirty: boolean
    hasBeenPersisted: boolean
    syncMessage: string
}
type SyncDB = {
    a: TestMeta & { name: string }
    b: TestMeta & { age: number }
}
type SyncDef<K extends keyof SyncDB> = {
    select: () => SelectQueryBuilder<SyncDB, K, {}>
    update: () => UpdateQueryBuilder<SyncDB, K, UpdateResult>
}

type NotSpecificEx = <R>(cont: (prop: SyncDef<keyof SyncDB>) => R) => R;

export function makeNotSpecificEx(property: SyncDef<keyof SyncDB>): NotSpecificEx {
    return <R>(cont: (prop: SyncDef<keyof SyncDB>) => R) => cont(property);
}

type Ex = <R>(cont: <K extends keyof SyncDB>(prop: SyncDef<K>) => R) => R;

export function makeEx<K extends keyof SyncDB>(property: SyncDef<K>): Ex {
    return <R>(cont: <K extends keyof SyncDB>(prop: SyncDef<K>) => R) => cont(property);
}

function testEx() {
    const exs = [
        makeEx(undefined! as SyncDef<"a">),
        makeEx(undefined! as SyncDef<"b">),
    ]
    exs.forEach(ex => {
        ex(ex => {
            ex.select().where("isDirty", "=", true);//error on true
            const res = ex.select().selectAll().execute();//returns Promise<({} & UnionToIntersection<SyncDB[K]>)[]>
        })
    })
    const nsexs = [
        makeNotSpecificEx(undefined! as SyncDef<"a">),
        makeNotSpecificEx(undefined! as SyncDef<"b">),
    ]
    nsexs.forEach(ex => {
        ex(ex => {
            const res = ex.select().where("isDirty", "=", true).selectAll().execute();//returns a union type
        })
    })
}
```

**Comment by koskimas** [comment]
When you write

```ts
    const exs = [
      makeEx(undefined! as SyncDef<'a'>),
      makeEx(undefined! as SyncDef<'b'>),
    ]
```

type of `exs` is just `Ex[]`. The actual types get erased.

**Comment by midwesterntechnology** [comment]
> When you write
> 
> ```ts
>     const exs = [
>       makeEx(undefined! as SyncDef<'a'>),
>       makeEx(undefined! as SyncDef<'b'>),
>     ]
> ```
> 
> type of `exs` is just `Ex[]`. The actual types get erased.

It's a continuation - if you loop over exs you get a function within which the types are inferred.

**Comment by koskimas** [comment]
Yep, that's what you are trying to do. Typescript disagrees.

**Comment by koskimas** [comment]
This is the correct type by the way `Promise<({} & UnionToIntersection<SyncDB[K]>)[]>`. If you try to substitute some specific `K` like `'a'` you get the correct type.

**Comment by midwesterntechnology** [comment]
> Yep, that's what you are trying to do. Typescript disagrees.
> 
> For example:
> 
> ```ts
> interface A {
>   a: string
> }
> 
> const as: A[] = [
>   { a: 'foo' },
>   { a: 'bar' }
> ]
> 
> function f(x: 'foo' | 'bar') {
> 
> }
> 
> f(as[0].a) // Error
> ```
> 
> This would work:
> 
> ```ts
> interface A {
>   a: string
> }
> 
> const as = [
>   { a: 'foo' },
>   { a: 'bar' }
> ] as const
> 
> function f(x: 'foo' | 'bar') {
> 
> }
> 
> f(as[0].a) // Works
> ```

A bit confused on this. If you're referring to the errors I just posted I wasn't expecting my previous example to work, it was just an example of the options I had and the issue with each.

In any case, just to be clear, it sounds like it is not going to be possible to infer these column types generically? Inherent limitations in typescript?

**Comment by koskimas** [comment]
Seems to be impossible using the kind of generics you are using, but I'm sure there's a way to do that in a type-safe way. Your examples are just so complex that it would take a long time for me to understand what you are after.

**Comment by midwesterntechnology** [comment]
Okay, let me try to put together a more concrete and simple example.

**Comment by midwesterntechnology** [comment]
Here is a very simple example which is really at the core of this. If I could do the following in a generic way, it would pretty much solve my issues.

```
import { AnyColumn, SelectQueryBuilder } from "kysely";

function getBuilder<K extends keyof DB>(): SelectQueryBuilder<DB, K, {}> {
    return undefined!;
}

type DB = {
    person: {
        id: number
    }
    pet: {
        id: string
    }
}

function getBy<K extends keyof DB>(column: AnyColumn<DB, K>, value: DB[K][typeof column]) {
    const builder = getBuilder<K>();
    return builder.where(column, "=", value);
}

getBy<"person">("id", 1);
```

**Comment by koskimas** [comment]
This works:

```ts
  function getBuilder<K extends keyof DB>(): SelectQueryBuilder<DB, K, {}> {
    return undefined!
  }

  type DB = {
    person: {
      id: number
    }
    pet: {
      id: string
    }
  }

  function getBy<K extends keyof DB, C extends keyof DB[K] & string>(
    column: C,
    value: FilterValueExpressionOrList<DB, K, C>
  ) {
    const builder = getBuilder<K>()
    return builder.where(column, '=', value)
  }

  getBy<'person', 'id'>('id', 1)
```

**Comment by midwesterntechnology** [comment]
Ok, starting to understand your approach. If I build the repos using my DB object I can encapsulate functionality there within a generic repo so the inferences are done upfront. Thanks!!!

I'd like to be able to use Kysely classes generically outside of my repos but that seems less likely to be possible. I will post a simple example of my attempt to do that shortly.

**Comment by midwesterntechnology** [comment]
I wound up building an abstractor type that allows me to build a partial view of the entity, essentially capturing all of the relevant inferences upfront on creation. It's a little cumbersome and would take a lot more work to cover the full query builder API but it does what I need it to do. Here's some basic code in case anyone ever has a similar problem.

```
type TestMeta = {
    id: string
    isDirty: boolean
    lastTriedToSync: Date
}

const dbInstance: Kysely<DB> = undefined!

type DB = {
    person: TestMeta & {
        id: string,
    }
}

type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (
    k: infer I
) => void
    ? I
    : never

type Abstractor<K extends keyof DB, AbstractedTo> = {
    select: SelectQueryBuilder<DB, K, {}>
    update: UpdateQueryBuilder<DB, K, {}>
    from: (row: UnionToIntersection<{} & DB[K]>) => AbstractedTo
    mutation: (row: Partial<AbstractedTo>) => MutationObject<DB, K>
    expressions: {
        [k in keyof AbstractedTo]: (value: AbstractedTo[k] | AbstractedTo[k][]) => [k, FilterValueExpressionOrList<DB, K, k>]
    }
}

function buildAbstractor<K extends keyof DB, A>(repo: Abstractor<K, A>) {
    return repo;
}

const personRepo = buildAbstractor<"person", TestMeta>({
    select: dbInstance.selectFrom("person"),
    update: dbInstance.updateTable("person"),
    mutation: row => row,
    from: row => row,
    expressions: {
        id: value => ["id", value],
        isDirty: value => ["isDirty", value],
        lastTriedToSync: value => ["lastTriedToSync", value],
    }
});

async function usingGenericAbstractor<K extends keyof DB>(repo: Abstractor<K, TestMeta>) {
    const [isDirtyCol, isDirtyValue] = repo.expressions.isDirty(true);
    const select = await repo.select.where(isDirtyCol, "=", isDirtyValue).selectAll().executeTakeFirstOrThrow();
    const converted = repo.from(select);
    const [idCol, idValue] = repo.expressions.id(converted.id);
    repo.update.set(repo.mutation({ isDirty: true })).where(idCol, "=", idValue);
}

```

**Comment by koskimas** [comment]
@midwesterntechnology I found a way to remove the `UnionToIntersection` type in 0.14.1. Some of the problems in the original code in this issue should no longer exist. Some of them still might.

**Comment by midwesterntechnology** [comment]
Nice! I'll check it out today.

**Comment by koskimas** [comment]
Actually it doesn't really seem to make a difference in your examples 😞 It may just be impossible to make your original example work.

**Comment by midwesterntechnology** [comment]
No problem, thanks for giving it a shot though! I will run through my old code and see if it fixes anything. For now I'm just using a code generator to build some generic repos.
